File: services/api/src/scrapers/source_health.rs
/// Public indexer source health tracking.
///
/// Mirrors workers/scrapers/source_health.py — records per-source success/timeout
/// outcomes in Redis and gates scraping when a source's success rate falls below
/// configured thresholds.
use std::collections::HashMap;

use fred::prelude::{HashesInterface, KeysInterface};

const METRICS_KEY_PREFIX: &str = "public_indexer_source_health:";
const DEFAULT_HEALTH_BUCKET: &str = "general";
const SCRAPER_METRICS_LATEST_KEY: &str = "scraper_metrics_latest:";
const SCRAPER_METRICS_HISTORY_KEY: &str = "scraper_metrics_history:";
const SCRAPER_METRICS_HISTORY_MAX: i64 = 100;
const SCRAPER_METRICS_TTL: i64 = 86400 * 7;

// ─── Config ───────────────────────────────────────────────────────────────────

/// Carries all health-gate config needed by the public-indexer scraper.
/// Cheap to clone (all primitives + one `fred::clients::Client` arc clone).
#[derive(Clone)]
pub struct HealthGateConfig {
    pub redis: fred::clients::Client,
    pub enabled: bool,
    pub min_samples: i64,
    pub min_success_rate: f64,
    pub max_timeout_rate: f64,
    pub counter_soft_cap: i64,
    pub decay_factor: f64,
    pub recovery_success_streak: i64,
    pub scope_mode: String,
    pub scope_override: String,
    pub metrics_ttl_seconds: i64,
}

// ─── Snapshot ────────────────────────────────────────────────────────────────

pub struct SourceHealthSnapshot {
    pub source_key: String,
    pub total: i64,
    pub success: i64,
    pub timeout: i64,
    pub challenge_solved: i64,
    pub consecutive_success: i64,
}

impl SourceHealthSnapshot {
    pub fn success_rate(&self) -> f64 {
        if self.total <= 0 {
            0.0
        } else {
            self.success as f64 / self.total as f64
        }
    }

    pub fn timeout_rate(&self) -> f64 {
        if self.total <= 0 {
            0.0
        } else {
            self.timeout as f64 / self.total as f64
        }
    }

    pub fn challenge_solve_rate(&self) -> f64 {
        if self.total <= 0 {
            0.0
        } else {
            self.challenge_solved as f64 / self.total as f64
        }
    }

    pub fn gate_status(
        &self,
        min_samples: i64,
        min_success_rate: f64,
        max_timeout_rate: f64,
    ) -> &'static str {
        if self.total < min_samples.max(1) {
            return "warming";
        }
        if self.success_rate() >= min_success_rate && self.timeout_rate() <= max_timeout_rate {
            "allowed"
        } else {
            "blocked"
        }
    }
}

// ─── Key helpers ─────────────────────────────────────────────────────────────

fn sanitize_scope_component(raw: &str) -> String {
    let lower = raw.trim().to_lowercase();
    let sanitized: String = lower
        .chars()
        .map(|c| {
            if c.is_alphanumeric() || c == '_' || c == '-' {
                c
            } else {
                '-'
            }
        })
        .collect();
    sanitized.trim_matches('-').to_string()
}

fn resolve_scope_key(scope_mode: &str, scope_override: &str) -> Option<String> {
    match scope_mode.trim().to_lowercase().as_str() {
        "global" | "" => None,
        "pod" => {
            let raw = std::env::var("PUBLIC_INDEXERS_SOURCE_HEALTH_SCOPE")
                .or_else(|_| std::env::var("POD_NAME"))
                .or_else(|_| std::env::var("HOSTNAME"))
                .unwrap_or_default();
            let s = sanitize_scope_component(&raw);
            Some(if s.is_empty() { "default".into() } else { s })
        }
        "custom" => {
            let s = sanitize_scope_component(scope_override);
            Some(if s.is_empty() { "default".into() } else { s })
        }
        _ => None,
    }
}

pub fn metrics_key(
    source_key: &str,
    health_bucket: &str,
    scope_mode: &str,
    scope_override: &str,
) -> String {
    let normalized_key = source_key.trim().to_lowercase();
    let bucket = {
        let b = sanitize_scope_component(health_bucket);
        if b.is_empty() {
            DEFAULT_HEALTH_BUCKET.to_string()
        } else {
            b
        }
    };
    match resolve_scope_key(scope_mode, scope_override) {
        None => format!("{METRICS_KEY_PREFIX}{bucket}:{normalized_key}"),
        Some(scope) => format!("{METRICS_KEY_PREFIX}{scope}:{bucket}:{normalized_key}"),
    }
}

// ─── Read / write ─────────────────────────────────────────────────────────────

pub async fn get_source_health(
    redis: &fred::clients::Client,
    source_key: &str,
    health_bucket: &str,
    scope_mode: &str,
    scope_override: &str,
) -> SourceHealthSnapshot {
    let key = metrics_key(source_key, health_bucket, scope_mode, scope_override);
    let raw: HashMap<String, String> = redis.hgetall(&key).await.unwrap_or_default();
    let get = |field: &str| -> i64 {
        raw.get(field)
            .and_then(|v| v.parse().ok())
            .unwrap_or(0)
    };
    if raw.is_empty() {
        return SourceHealthSnapshot {
            source_key: source_key.to_string(),
            total: 0,
            success: 0,
            timeout: 0,
            challenge_solved: 0,
            consecutive_success: 0,
        };
    }
    SourceHealthSnapshot {
        source_key: source_key.to_string(),
        total: get("total"),
        success: get("success"),
        timeout: get("timeout"),
        challenge_solved: get("challenge_solved"),
        consecutive_success: get("consecutive_success"),
    }
}

pub async fn record_source_outcome(
    redis: &fred::clients::Client,
    source_key: &str,
    success: bool,
    timed_out: bool,
    challenge_solved: bool,
    health_bucket: &str,
    scope_mode: &str,
    scope_override: &str,
    soft_cap: i64,
    decay_factor: f64,
    ttl_seconds: i64,
) {
    let key = metrics_key(source_key, health_bucket, scope_mode, scope_override);
    let total: i64 = redis
        .hincrby::<i64, _, _>(&key, "total", 1)
        .await
        .unwrap_or(0);
    if success {
        let _: Result<i64, _> = redis.hincrby(&key, "success", 1).await;
    }
    if timed_out {
        let _: Result<i64, _> = redis.hincrby(&key, "timeout", 1).await;
    }
    if challenge_solved {
        let _: Result<i64, _> = redis.hincrby(&key, "challenge_solved", 1).await;
    }
    if success && !timed_out {
        let _: Result<i64, _> = redis.hincrby(&key, "consecutive_success", 1).await;
    } else {
        let mut fields = HashMap::new();
        fields.insert("consecutive_success".to_string(), "0".to_string());
        let _: Result<i64, _> = redis.hset(&key, fields).await;
    }

    if total >= soft_cap {
        decay_source_counters(redis, &key, decay_factor).await;
    }

    let _: Result<bool, _> = redis.expire(&key, ttl_seconds, None).await;
}

async fn decay_source_counters(
    redis: &fred::clients::Client,
    key: &str,
    decay_factor: f64,
) {
    let raw: HashMap<String, String> = match redis.hgetall(key).await {
        Ok(m) => m,
        Err(_) => return,
    };
    if raw.is_empty() {
        return;
    }
    let get = |field: &str| -> i64 {
        raw.get(field)
            .and_then(|v| v.parse().ok())
            .unwrap_or(0)
    };
    let current_total = get("total");
    if current_total <= 0 {
        return;
    }
    let decayed_total = ((current_total as f64 * decay_factor).max(1.0)) as i64;
    let decayed_success =
        ((get("success") as f64 * decay_factor) as i64).min(decayed_total).max(0);
    let decayed_timeout =
        ((get("timeout") as f64 * decay_factor) as i64).min(decayed_total).max(0);
    let decayed_challenge =
        ((get("challenge_solved") as f64 * decay_factor) as i64).min(decayed_total).max(0);
    let decayed_streak = get("consecutive_success").min(decayed_total).max(0);

    let mut fields = HashMap::new();
    fields.insert("total".to_string(), decayed_total.to_string());
    fields.insert("success".to_string(), decayed_success.to_string());
    fields.insert("timeout".to_string(), decayed_timeout.to_string());
    fields.insert("challenge_solved".to_string(), decayed_challenge.to_string());
    fields.insert("consecutive_success".to_string(), decayed_streak.to_string());
    let _: Result<i64, _> = redis.hset(key, fields).await;
}

pub async fn is_source_within_budget(
    redis: &fred::clients::Client,
    source_key: &str,
    min_samples: i64,
    min_success_rate: f64,
    max_timeout_rate: f64,
    health_bucket: &str,
    scope_mode: &str,
    scope_override: &str,
) -> bool {
    let snapshot = get_source_health(redis, source_key, health_bucket, scope_mode, scope_override).await;
    if snapshot.total < min_samples.max(1) {
        return true; // not enough samples yet → allow
    }
    snapshot.success_rate() >= min_success_rate && snapshot.timeout_rate() <= max_timeout_rate
}

// ─── Scraper run metrics ──────────────────────────────────────────────────────

/// Save a minimal scraper-run metrics record to Redis so the admin dashboard
/// "Recent Media Search Runs" page shows Rust-server scrape activity.
///
/// Matches the JSON schema that Python's `ScraperMetrics.save_to_redis()` writes.
pub async fn save_scraper_run_metrics(
    redis: &fred::clients::Client,
    scraper_name: &str,
    imdb_id: Option<&str>,
    title: &str,
    season: Option<i32>,
    episode: Option<i32>,
    found: usize,
    processed: usize,
    skipped: usize,
    errors: usize,
    skip_reasons: &HashMap<String, usize>,
    start_ts: &chrono::DateTime<chrono::Utc>,
    end_ts: &chrono::DateTime<chrono::Utc>,
) {
    use fred::prelude::ListInterface;

    let duration_secs = (*end_ts - *start_ts).num_milliseconds() as f64 / 1000.0;

    let mut skip_reasons_json = serde_json::Map::new();
    for (k, v) in skip_reasons {
        skip_reasons_json.insert(k.clone(), serde_json::json!(v));
    }

    // Build the formatted summary matching Python's format_summary()
    let skips_str: String = if skip_reasons.is_empty() {
        "N/A".to_string()
    } else {
        skip_reasons
            .iter()
            .map(|(k, v)| format!("{k}:{v}"))
            .collect::<Vec<_>>()
            .join(", ")
    };

    let formatted_summary = format!(
        "\n{sep}\n{title_line}\n{sep}\n\nMeta ID: {meta_id}\nTitle: {title}\n\nDuration: {dur:.2} seconds\n\nItems:\n  ├─ Found     : {found}\n  ├─ Processed : {processed}\n  ├─ Skipped   : {skipped}\n  └─ Errors    : {errors}\n\nSkip Reasons: {skips}\n{sep}\n",
        sep = "=".repeat(80),
        title_line = format!("{} Scraping Metrics Summary", scraper_name.to_uppercase()).as_str().chars().collect::<String>(),
        meta_id = imdb_id.unwrap_or("unknown"),
        title = title,
        dur = duration_secs,
        found = found,
        processed = processed,
        skipped = skipped,
        errors = errors,
        skips = skips_str,
    );

    let summary = serde_json::json!({
        "scraper_name": scraper_name,
        "timestamp": start_ts.format("%Y-%m-%dT%H:%M:%S%.6f").to_string(),
        "end_timestamp": end_ts.format("%Y-%m-%dT%H:%M:%S%.6f").to_string(),
        "duration_seconds": duration_secs,
        "meta_id": imdb_id,
        "meta_title": title,
        "season": season,
        "episode": episode,
        "skip_scraping": false,
        "total_items": {
            "found": found,
            "processed": processed,
            "skipped": skipped,
            "errors": errors
        },
        "error_counts": {},
        "skip_reasons": skip_reasons_json,
        "quality_distribution": {},
        "source_distribution": {},
        "indexer_stats": {},
        "formatted_summary": formatted_summary,
    });

    let Ok(json_str) = serde_json::to_string(&summary) else {
        return;
    };

    // Save latest
    let latest_key = format!("{SCRAPER_METRICS_LATEST_KEY}{scraper_name}");
    let _: Result<(), _> = redis
        .set(&latest_key, json_str.clone(), Some(fred::types::Expiration::EX(SCRAPER_METRICS_TTL)), None, false)
        .await;

    // Prepend to history list, cap at SCRAPER_METRICS_HISTORY_MAX
    let history_key = format!("{SCRAPER_METRICS_HISTORY_KEY}{scraper_name}");
    let _: Result<i64, _> = redis.lpush(&history_key, json_str).await;
    let _: Result<(), _> = redis
        .ltrim(&history_key, 0, SCRAPER_METRICS_HISTORY_MAX - 1)
        .await;
    let _: Result<bool, _> = redis.expire(&history_key, SCRAPER_METRICS_TTL, None).await;
}

File: python-deprecated/workers/scrapers/source_health.py
from __future__ import annotations

import os
import re
from dataclasses import dataclass

from db.config import settings
from db.redis_database import REDIS_ASYNC_CLIENT

METRICS_KEY_PREFIX = "public_indexer_source_health:"
METRICS_TTL_SECONDS = settings.public_indexers_source_health_metrics_ttl_seconds
DEFAULT_HEALTH_BUCKET = "general"


@dataclass(frozen=True)
class SourceHealthScope:
    mode: str
    scope_key: str | None


@dataclass(frozen=True)
class SourceHealthSnapshot:
    source_key: str
    total: int
    success: int
    timeout: int
    challenge_solved: int
    consecutive_success: int

    @property
    def success_rate(self) -> float:
        if self.total <= 0:
            return 0.0
        return self.success / self.total

    @property
    def timeout_rate(self) -> float:
        if self.total <= 0:
            return 0.0
        return self.timeout / self.total

    @property
    def challenge_solve_rate(self) -> float:
        if self.total <= 0:
            return 0.0
        return self.challenge_solved / self.total


def _metrics_key(source_key: str, health_bucket: str = DEFAULT_HEALTH_BUCKET) -> str:
    normalized_source_key = (source_key or "").strip().lower()
    normalized_bucket = _sanitize_scope_component(health_bucket) or DEFAULT_HEALTH_BUCKET
    scope = _resolve_scope()
    if not scope.scope_key:
        return f"{METRICS_KEY_PREFIX}{normalized_bucket}:{normalized_source_key}"
    return f"{METRICS_KEY_PREFIX}{scope.scope_key}:{normalized_bucket}:{normalized_source_key}"


def _sanitize_scope_component(raw_value: str | None) -> str:
    normalized = re.sub(r"[^a-z0-9_-]+", "-", str(raw_value or "").strip().lower()).strip("-")
    return normalized


def _resolve_scope() -> SourceHealthScope:
    mode = (settings.public_indexers_source_health_scope_mode or "global").strip().lower()
    if mode not in {"global", "pod", "custom"}:
        mode = "global"
    if mode == "global":
        return SourceHealthScope(mode=mode, scope_key=None)

    raw_scope = ""
    if mode == "pod":
        raw_scope = (
            os.getenv("PUBLIC_INDEXERS_SOURCE_HEALTH_SCOPE") or os.getenv("POD_NAME") or os.getenv("HOSTNAME") or ""
        )
    elif mode == "custom":
        raw_scope = settings.public_indexers_source_health_scope

    sanitized_scope = _sanitize_scope_component(raw_scope)
    if not sanitized_scope:
        return SourceHealthScope(mode=mode, scope_key="default")
    return SourceHealthScope(mode=mode, scope_key=sanitized_scope)


def get_source_health_scope() -> SourceHealthScope:
    return _resolve_scope()


def _parse_counter(raw_value) -> int:
    if raw_value is None:
        return 0
    if isinstance(raw_value, bytes):
        raw_value = raw_value.decode("utf-8", errors="ignore")
    try:
        return int(raw_value)
    except (TypeError, ValueError):
        return 0


def _get_counter(raw: dict, field: str) -> int:
    if not raw:
        return 0
    return _parse_counter(raw.get(field) or raw.get(field.encode("utf-8")))


async def _decay_source_counters(key: str) -> None:
    raw = await REDIS_ASYNC_CLIENT.hgetall(key)
    if not raw:
        return

    current_total = _get_counter(raw, "total")
    if current_total <= 0:
        return

    decay_factor = settings.public_indexers_source_health_decay_factor
    decayed_total = max(1, int(current_total * decay_factor))
    decayed_success = min(decayed_total, max(0, int(_get_counter(raw, "success") * decay_factor)))
    decayed_timeout = min(decayed_total, max(0, int(_get_counter(raw, "timeout") * decay_factor)))
    decayed_challenge = min(decayed_total, max(0, int(_get_counter(raw, "challenge_solved") * decay_factor)))
    current_streak = _get_counter(raw, "consecutive_success")
    decayed_streak = min(max(0, current_streak), decayed_total)

    await REDIS_ASYNC_CLIENT.hset(
        key,
        mapping={
            "total": decayed_total,
            "success": decayed_success,
            "timeout": decayed_timeout,
            "challenge_solved": decayed_challenge,
            "consecutive_success": decayed_streak,
        },
    )


async def record_source_outcome(
    source_key: str,
    *,
    success: bool,
    timed_out: bool = False,
    challenge_solved: bool = False,
    health_bucket: str = DEFAULT_HEALTH_BUCKET,
) -> None:
    key = _metrics_key(source_key, health_bucket)
    total = await REDIS_ASYNC_CLIENT.hincrby(key, "total", 1)
    if success:
        await REDIS_ASYNC_CLIENT.hincrby(key, "success", 1)
    if timed_out:
        await REDIS_ASYNC_CLIENT.hincrby(key, "timeout", 1)
    if challenge_solved:
        await REDIS_ASYNC_CLIENT.hincrby(key, "challenge_solved", 1)

    clean_success = success and not timed_out
    if clean_success:
        await REDIS_ASYNC_CLIENT.hincrby(key, "consecutive_success", 1)
    else:
        await REDIS_ASYNC_CLIENT.hset(key, "consecutive_success", 0)

    if total >= settings.public_indexers_source_health_counter_soft_cap:
        await _decay_source_counters(key)

    await REDIS_ASYNC_CLIENT.expire(key, METRICS_TTL_SECONDS)


async def get_source_health(
    source_key: str,
    *,
    health_bucket: str = DEFAULT_HEALTH_BUCKET,
) -> SourceHealthSnapshot:
    key = _metrics_key(source_key, health_bucket)
    raw = await REDIS_ASYNC_CLIENT.hgetall(key)
    if not raw:
        return SourceHealthSnapshot(
            source_key=source_key,
            total=0,
            success=0,
            timeout=0,
            challenge_solved=0,
            consecutive_success=0,
        )
    return SourceHealthSnapshot(
        source_key=source_key,
        total=_get_counter(raw, "total"),
        success=_get_counter(raw, "success"),
        timeout=_get_counter(raw, "timeout"),
        challenge_solved=_get_counter(raw, "challenge_solved"),
        consecutive_success=_get_counter(raw, "consecutive_success"),
    )


async def is_source_within_budget(
    source_key: str,
    *,
    min_samples: int,
    min_success_rate: float,
    max_timeout_rate: float,
    health_bucket: str = DEFAULT_HEALTH_BUCKET,
) -> bool:
    snapshot = await get_source_health(source_key, health_bucket=health_bucket)
    if snapshot.total < max(1, min_samples):
        return True
    return snapshot.success_rate >= min_success_rate and snapshot.timeout_rate <= max_timeout_rate
